### ISSUE TITLE
Modernising to single package with esm and types

### ISSUE THREAD

**Issue by daKmoR** [joaonuno/tree-model-js#84]
hey there 👋

I was wondering if there would be any interest in "updating" the code to esm?
I'm also very conscious of the number of dependencies and I think we could get that down to 0?
I also noticed some typing issues for children... 

anyways - I could do all the work and do esm, 0 dependencies, and improving types... would you like to release it maybe as the next major version?
or would you prefer me to fork?

**Comment by joaonuno** [comment]
Hey! I like your suggestions and would appreciate your contributions. I can't commit much time to this so please break it down in separate MRs to make it more digestible.
I'm not familiar with how much breaking is changing the distributed module type but I guess a new major makes sense.
Regarding number of dependencies, I'm not strict on that, actually the two dependencies are also mine but I agree we should get rid of them because:
- [merge-sort-js](https://github.com/joaonuno/merge-sort-js) was built just to provide stable sorting (when the lib was developed the JS sort method wasn't stable but I think it is now both in node and browsers, please confirm this)
- [find-insert-index-js](https://github.com/joaonuno/find-insert-index-js) is doing a linear search when it could do a binary search (not sure if there's anything native to replace this)

I'm not a TS user so I don't know what's missing or needs improvements in regard with typing. If it's possible to make those improvements separately and released as a new patch version that would be desirable.

I also ask you to avoid changing code and tests in the same MR to keep it more controlled in terms of breaking changes.

**Comment by daKmoR** [comment]
sounds good 👍

First PR will be about moving to ESM and classes which should mostly be syntactical changes while hardly touching logical code. Just to be sure this will raise the requirement to node 14+ (current active LTS)

we can then align on how to take the next steps 🤗

**Comment by joaonuno** [comment]
> Just to be sure this will raise the requirement to node 14+ (current active LTS)

Did you check if Node 12 (still receiving [security updates](https://endoflife.date/nodejs)) already supports ESM? If not, or if it's incomplete, then do bump the minimum required to 14.

**Comment by daKmoR** [comment]
Suggested steps

- [x] move to esm  ([PR](https://github.com/joaonuno/tree-model-js/pull/85))
- [x] keep working in a branch? (do next PRs to it? there are more changes needed)
- [x] format with prettier
- [ ] add github action to automatically run tests in PRs
- [ ] split into `src/TreeModel.js`, `src/Node.js`, `src/walkStrategies.js` ([PR](https://github.com/joaonuno/tree-model-js/pull/90))
- [x] remove dtslint, jshint
- [x] update eslint
- [ ] evaluate what would be needed to become a package with 0 dependencies
- [ ] add typescript and auto-generate types based on jsdoc

**Comment by daKmoR** [comment]
Node 12 indeed also supports esm but IIRC it does not support export maps which I would also like to add

**Comment by daKmoR** [comment]
@joaonuno thx for the approval 🤗

how would you like to proceed? 

suggestion:
1. merge the open PR into another branch like `next`
1. you would need to create and push this `next` branch before I can choose it as a target
1. all following merge requests would target `next`
1. ... work work work
1. after all steps are done merge `next` into master and released it as 2.0.0 (potentially doing some pre-releases before)

**Comment by joaonuno** [comment]
Thanks for the suggestion. Just created `next` branch.

**Comment by daKmoR** [comment]
hey there 👋
any chance you can merge https://github.com/joaonuno/tree-model-js/pull/85 so I can make further PRs on the `next` branch?

**Comment by joaonuno** [comment]
Done, sorry for the delay.

**Comment by daKmoR** [comment]
I got pulled in some other tasks 🙈  but I'm back... finished "remove dtslint, jshint" and "update eslint" via https://github.com/joaonuno/tree-model-js/pull/89

I brought those tasks forward as while splitting the files I noted I didn't get any hints about missing imports or functions... so I want that fixed first... as it will make the work easier

**Comment by joaonuno** [comment]
Hi,
Was wondering if, when you mentioned that you're planning some type improvements, you also noticed the potential improvement mentioned in #88. If so, maybe it's something that can enter in the next major. WDYT?

**Comment by daKmoR** [comment]
yes - I assume this will come up once we add type checking for js files

just finished splitting it into multiple smaller parts in https://github.com/joaonuno/tree-model-js/pull/90

I think these are the bigger remaining open tasks
- [x] evaluate what would be needed to become a package with 0 dependencies
- [ ] add typescript and auto-generate types based on jsdoc
- [x] Make it type module and rename all mjs to js
- [ ] Beta Release preparation (export map, readme, changelog, ...)

**Comment by daKmoR** [comment]
can I ask you to review/merge #90? 🙏

**Comment by daKmoR** [comment]
here we are - 0 dependencies 💪  see #91

**Comment by daKmoR** [comment]
yeah, huge changes in 
- https://github.com/joaonuno/tree-model-js/pull/93

what do you think? how do you want to proceed?
